File: app/src/main/java/cz/anty/purkynka/MainData.java
package cz.anty.purkynka;

import android.content.Context;
import android.content.SharedPreferences;

import eu.codetopic.utils.data.getter.DataGetter;
import eu.codetopic.utils.data.preferences.provider.BasicSharedPreferencesProvider;
import eu.codetopic.utils.data.preferences.support.PreferencesGetterAbs;
import eu.codetopic.utils.data.preferences.PreferencesData;

import static cz.anty.purkynka.PrefNames.*;

/**
 * Created by anty on 6/20/17.
 *
 * @author anty
 */
public class MainData extends PreferencesData<SharedPreferences> {

    public static final DataGetter<MainData> getter = new Getter();

    private static final String LOG_TAG = "MainData";
    private static final int SAVE_VERSION = 0;

    private static MainData mInstance = null;

    private MainData(Context context) {
        super(context, new BasicSharedPreferencesProvider(context,
                FILE_NAME_MAIN_DATA, Context.MODE_PRIVATE), SAVE_VERSION);
    }

    public static void initialize(Context context) {
        if (mInstance != null) throw new IllegalStateException(LOG_TAG + " is still initialized");
        mInstance = new MainData(context);
        mInstance.init();
    }

    private static final class Getter extends PreferencesGetterAbs<MainData> {

        @Override
        public MainData get() {
            return mInstance;
        }

        @Override
        public Class<MainData> getDataClass() {
            return MainData.class;
        }
    }
}

File: app/src/main/java/cz/anty/purkynka/MainData.kt
/*
 * app
 * Copyright (C)   2017  anty
 *
 * This program is free  software: you can redistribute it and/or modify
 * it under the terms  of the GNU General Public License as published by
 * the Free Software  Foundation, either version 3 of the License, or
 * (at your option) any  later version.
 *
 * This program is distributed in the hope that it  will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied  warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
 * GNU General Public License for more details.
 *
 * You  should have received a copy of the GNU General Public License
 * along  with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package cz.anty.purkynka

import android.content.Context
import android.content.SharedPreferences

import eu.codetopic.utils.data.preferences.VersionedPreferencesData
import eu.codetopic.utils.data.preferences.provider.BasicSharedPreferencesProvider
import eu.codetopic.utils.data.preferences.support.PreferencesGetterAbs

import cz.anty.purkynka.utils.*
import eu.codetopic.utils.data.preferences.support.PreferencesCompanionObject

/**
 * @author anty
 */
class MainData private constructor(context: Context) :
        VersionedPreferencesData<SharedPreferences>(context,
                BasicSharedPreferencesProvider(context, FILE_NAME_MAIN_DATA),
                SAVE_VERSION) {

    companion object : PreferencesCompanionObject<MainData>(MainData.LOG_TAG, ::MainData, ::Getter) {

        private const val LOG_TAG = "MainData"
        private const val SAVE_VERSION = 0
    }

    @Synchronized
    override fun onUpgrade(editor: SharedPreferences.Editor, from: Int, to: Int) {
        when (from) {
            -1 -> {
                // First start, nothing to do
            } // No more versions yet
        }
    }

    private class Getter : PreferencesGetterAbs<MainData>() {

        override fun get() = instance

        override val dataClass: Class<out MainData>
            get() = MainData::class.java
    }
}
